### PR TITLE
Remove filter setters.

### LINKED ISSUE #185: Filtered Ordered Associations Shouldn't be Settable
I ran into this while working on moving the new AF::Aggregations work over here.

What does it mean to set `#objects`, or append to `#objects`? The way it worked before was "append anything to the end of the list", but that does unexpected things to the order as a whole - because ordering is at the member level, not at the PCDM::Object level. For instance:

```ruby
col1 = Collection.new
col2 = Collection.new
m1 = Member.new
m2 = Member.new
col1.ordered_members += [m1, col2, m2]
# Reorder objects
col1.objects # => [m1, m2]
col1.objects = [m2, m1]
col1.ordered_members [col2, m2, m1] # => This is definitely wrong, but it's easy to happen, because we're ordering out of context. Things get even worse when you start trying to delete.
```

My proposal is this: Only #members, #ordered_members, and #ordered_member_proxies should be settable.

### PR DESCRIPTION
**Pull request by terrellt**
Closes #185.

This is a **HUGE** backwards incompatible change that will break anything that tries to depend on it, but it never made sense for filtered associations to be settable when implying order. The other option is to leave the filtered association, but assume that using them means they will not be added to the ordered list.

### CODE REVIEW

**Comment by terrellt** [comment]
@elrayle Could you comment on this?

**Comment by mjgiarlo** [comment]
:+1: to having @elrayle review this

File: spec/hydra/pcdm/models/file_spec.rb [hunk 1]
```diff
@@ -50,7 +50,7 @@
       # complains that it's a server managed property. This test
       # is mostly to document this situation.
       file.file_hash = 'the-hash'
-      expect { file.save }.to raise_error(Ldp::Conflict, %r{Could not remove triple containing predicate http://www.loc.gov/premis/rdf/v1#hasMessageDigest to node .*})
+      expect { file.save }.to raise_error(Ldp::BadRequest, %r{Could not persist triple containing predicate http://www.loc.gov/premis/rdf/v1#hasMessageDigest to node .*})
```
Comment (mjgiarlo) [inline spec/hydra/pcdm/models/file_spec.rb]: Can you say something about this change?

**Comment by mjgiarlo** [comment]
Why is `#collection_ids` coerced to an array and not `#object_ids`? I ask because the code is littered with `object_ids.to_a` which might warrant doing that in the method itself.

**Comment by mjgiarlo** [comment]
I'm also struggling with the 5.5% reduction in test coverage ;)

**Comment by terrellt** [comment]
> I'm also struggling with the 5.5% reduction in test coverage ;)

:+1: I'll take a look at that.

File: spec/hydra/pcdm/models/file_spec.rb (hunk 1 above)
Comment (terrellt) [inline spec/hydra/pcdm/models/file_spec.rb]: Changing it back - that's how it fails locally.

**Comment by terrellt** [comment]
> Why is #collection_ids coerced to an array and not #object_ids? I ask because the code is littered with object_ids.to_a which might warrant doing that in the method itself.

:+1: The to_a was leftover when I was playing with #lazy

File: spec/hydra/pcdm/models/file_spec.rb (hunk 1 above)
Comment (jcoyne) [inline spec/hydra/pcdm/models/file_spec.rb]: I think this changed between Fedora 4.2 and 4.3

**Comment by mjgiarlo** [comment]
Coverage is much improved! :clap:

**Comment by mjgiarlo** [comment]
>> Why is #collection_ids coerced to an array and not #object_ids? I ask because the code is littered with object_ids.to_a which might warrant doing that in the method itself.
> :+1: The to_a was leftover when I was playing with #lazy

So, the code is no longer littered with `object_ids.to_a`, and that's great. I'm not clear on why `#collection_ids` uses `#to_a` and `#object_ids` does not, though.

**Comment by elrayle** [comment]
There are two questions here...

1) Why are filters being removed?  They were added to ensure data integrity by enforcing the model constraints.  And to provide a common sense API that allows users of the model to think of groups of collections and groups of objects even though under the covers we store them all in members.

BTW, this is a very major change from PCDM as it was demonstrated at Hydra Connect.  The Dive in to Hydra-Works tutorial uses filter methods throughout.

2) If filters are going to be removed, are the code changes sufficient?  I brought the code up in a console and found several issues.  I did not fully test.  I just wanted to get a sense of how things changed.

Try this...
```
> require 'hydra/pcdm'
> col1 = Hydra::PCDM::Collection.new
> col2 = Hydra::PCDM::Collection.new
> obj1 = Hydra::PCDM::Object.new

> col1.collections << col2   # => [#<Hydra::PCDM::Collection id: nil, head_id: nil, tail_id: nil>]
> col1.collections           # => []
> col1.objects << obj1       # => [#<Hydra::PCDM::Collection id: nil, head_id: nil, tail_id: nil>]
> col1.objects               # => []

> col2.in_collections        # ERROR because pcdm_collection? is defined by the filter
NoMethodError: undefined method `pcdm_collection?' for #<ActiveFedora::Base id: "col-1">
from /Users/elr37/.rvm/gems/ruby-2.1.2/gems/activemodel-4.2.4/lib/active_model/attribute_methods.rb:433:in `method_missing'

> obj.in_collections         # ERROR because pcdm_collection? is defined by the filter
NoMethodError: undefined method `pcdm_collection?' for #<ActiveFedora::Base id: "col-1">
from /Users/elr37/.rvm/gems/ruby-2.1.2/gems/activemodel-4.2.4/lib/active_model/attribute_methods.rb:433:in `method_missing'
```

I assume, but did not test the same methods for objects.

NOTE: The test for in_collections passes because of the mocking.  Prior to the first allow-receive-and_return, the error message matches that seen above.  After the first one, in_collections returns [].  After the second, it returns the parent collections.  So the mocking is hiding the errors seen in the ad hoc testing.

**Comment by jcoyne** [comment]
@elrayle the filters are not being removed. The reader methods will still be there. This is what a filter ought to be, a constraint on the universe that returns a subset.  @terrellt  is just removing the setter methods for the filter.   The setters on a filter make it very hard to understand the ordering of the universe.  See #185 for some very good examples of that problem.

**Comment by elrayle** [comment]
@jcoyne That makes more sense.  I did some more ad hoc testing under this new perspective and here are some failures that need to be covered by the spec tests.

```
require 'hydra/pcdm'

# setup
col1 = Hydra::PCDM::Collection.new
col2 = Hydra::PCDM::Collection.new
obj1 = Hydra::PCDM::Object.new
obj2 = Hydra::PCDM::Object.new

# ERROR old setters should throw missing method or some other error
col1.collections << col2  # fails silently
col1.objects << obj1      # fails silently
col1.members              # returns []

obj1.objects << obj2      # fails silently
obj1.members              # returns []

# now add members the correct way
col1.members << col2
col1.members << obj1
obj1.members << obj2

# correctly get by type
col1.collections         # [col2]
col1.objects             # [obj1]

obj1.objects             # [obj2]

# ERROR get parent collections gets NoMethodError
col2.in_collections   # 'NoMethodError: undefined method `pcdm_collection?' for #<ActiveFedora::Base id: "col-1">'
obj1.in_collections   # 'NoMethodError: undefined method `pcdm_collection?' for #<ActiveFedora::Base id: "col-1">'

obj2.in_objects       # 'NoMethodError: undefined method `pcdm_object?' for #<ActiveFedora::Base id: "col-1">'

col1.pcdm_collection?  # => true
obj1.pcdm_object?    # => true
```

NOTE: Previously I stated that pcdm_collection? was defined by the filter association.  That is incorrect.  That method is defined in the model concerns.  I'm not sure why the in_collections and in_objects methods are failing to find those methods.

**Comment by jcoyne** [comment]
@elrayle these are not failing:
```
col1.collections << col2  # fails silently
col1.objects << obj1      # fails silently
```
The methods `collections` and `objects` are just returning a new array each time, so while you are free to modify the returned value, it's not going to return the same object on subsequent calls.

**Comment by elrayle** [comment]
@jcoyne I see how the code is functioning as it is written now that the filter association is removed.  From the user's perspective this represents a major change.

What is the plan for releasing this?  We have already presented PCDM to the community with the filters.  It will be challenging to make deprecated methods for the code that no longer works as before since the removed filters previously used reflection to add methods <<, +=, and =.

Ideally, there would be a period of deprecation for the following methods...
```
col1.collections = [col2]
col1.collections << col3
col1.collections += [col4]

col1.objects = [obj1]
col1.objects << obj2
col1.objects += [obj3]


obj1.objects = [obj2]
obj1.objects << obj3
obj1.objects += [obj4]
```

This will likely break code at all levels going up the stack from Hydra::Works - Curation Concerns - Sufia.  And now, there may be others outside our group that are using the code following the above API which was documented and presented at Hydra Connect.  The API documentation and Dive in to Hydra-Works tutorials need to be updated with this release.  Is there a ticket for making the same changes in Hydra-Works?

**Comment by terrellt** [comment]
> This will likely break code at all levels going up the stack from Hydra::Works - Curation Concerns - Sufia

Yup. Well, Sufia won't break yet because that sprint hasn't started (has it?), but everything else is true. I intended to fix it going up the stack after this was merged.

I'm not sure why in_collections/in_objects is failing. If I get rid of the stubbing in the tests, those pass. Maybe see what #aggregated_by returns?

Deprecation warnings - I could get them in, but it would be a mess of code that did nothing except make the upgrade path obvious. I'll leave whether or not to do so to the community, but considering we're not even at 1.0 yet (not to mention in -labs), I'm concerned that the technical debt wouldn't be worth it.

**Comment by elrayle** [comment]
@terrellt I feel your pain in not wanting to take the time to add code that will be ripped out later, especially if it takes significant time to accomplish this.  If others agree that, in this case, it is ok to skip deprecation, then I would not oppose that.  Although, as a general rule, I don't think we should remove or change methods without providing deprecation warnings.  There is nothing worse than working higher in the stack and something changes out from under you without warning or a clear path for fixing the issues.

My suggestion would be...
* Announce the change and ask for comments on hydra-tech.  Provide a clear statement that the API is changing and give an example of the impact on code following the pattern `col1.collections << col2`, etc. including that this should change to `col1.members << col2`, etc.
* Minimally coordinate the commit at hydra-pcdm and hydra-works so they happen within a very short time span of each other
* Update the API doc and tutorials and coordinate the commit of these changes with the commit of the code  (Make a clear note of the API change in the API doc and tutorial to avoid confusion for those returning to the tutorial.)

**Comment by terrellt** [comment]
So for me this was a step in a process that looked like this:

1. Merge this.
2. Fix up Hydra::Works
3. Fix up CurationConcerns
4. Switch over Hydra::PCDM to new ordering API in Aggregations (couldn't do it with these filter    mutators)
5. Fix up Hydra::Works
6. Fix up CurationConcerns

It sounds like the process now would be:

1. Post to hydra-tech about this API change
2. Merge this
3. Fix up Hydra::Works
4. Fix up CurationConcerns
5. Post to hydra-tech about another API change from switching to new ordering API in Aggregations
6. Send Hydra::PCDM PR using that API
7. Fix up Hydra::Works
8. Fix up CurationConcerns

Yes?

If so, I might have to put off getting the new ordering API implemented, as we have local work to focus on in the next few weeks. I should be able to get 1-4 done in relatively short order, but the rest will have to wait, which means that we'll have to add another step which is "fix up Sufia" (since that work will be going on in the meantime.)

**Comment by jcoyne** [comment]
@elrayle I don't think we can deprecate methods that are on the Array class. e.g.: 

```
col1.collections << col3
obj1.objects << obj2
```

**Comment by jcoyne** [comment]
I also don't feel strongly about deprecations considering this is 0.x version and in projecthydra-labs.  At this stage of development we should refactor mercilessly to get things right.  Obviously this will be painful, but better to make a painful change now than when more people are using this.

**Comment by terrellt** [comment]
Posted on hydra-tech at https://groups.google.com/forum/#!topic/hydra-tech/NQfqQxRKNDg

**Comment by mjgiarlo** [comment]
Middle-road proposal, so that more can get done sooner (and taking into account that this is a 0.x stack, this is -labs, and most if not all of the users of this stack are very likely on this thread):

1. Notify hydra-tech that multiple API changes are coming to this stack
2. Merge this
3. Fix Hydra::Works
4. Fix CurationConcerns
6. Switch over Hydra::PCDM to new ordering API in Aggregations
7. Fix Hydra::Works
8. Fix CurationConcerns
5. Fix tutorials
6. Fix README documentation.

**Comment by terrellt** [comment]
@mjgiarlo :+1: That seems like the reasonable course here.

**Comment by terrellt** [comment]
As an aside, removed the explicit to_a on collection_ids.

**Comment by elrayle** [comment]
@mjgiarlo :+1:

**Comment by elrayle** [comment]
There is still one question unresolved for me.  Why do the tests for #in_collections pass, but the following fails...

```
require 'hydra/pcdm'

col1 = Hydra::PCDM::Collection.new
col2 = Hydra::PCDM::Collection.new
col1.members << col2

col2.in_collections   # 'NoMethodError: undefined method `pcdm_collection?' for #<ActiveFedora::Base id: "col-1">'
```

**Comment by terrellt** [comment]
@elrayle Because col1 isn't saved.

**Comment by terrellt** [comment]
@elrayle 

```ruby
Resolving dependencies...
irb(main):001:0> col1 = Hydra::PCDM::Collection.new
=> #<Hydra::PCDM::Collection id: nil, head_id: nil, tail_id: nil>
irb(main):002:0> col2 = Hydra::PCDM::Collection.new
=> #<Hydra::PCDM::Collection id: nil, head_id: nil, tail_id: nil>
irb(main):003:0> col1.members << col2
=> [#<Hydra::PCDM::Collection id: nil, head_id: nil, tail_id: nil>]
irb(main):004:0> col1.save
=> true
irb(main):005:0> col2.in_collections
=> [#<Hydra::PCDM::Collection id: "26/8a/66/03/268a6603-990f-417b-b039-dcc74cf5fde1", head_id: "26/8a/66/03/268a6603-990f-417b-b039-dcc74cf5fde1/members/77fcda6e-0504-42f7-b430-524f36a88991", tail_id: "26/8a/66/03/268a6603-990f-417b-b039-dcc74cf5fde1/members/77fcda6e-0504-42f7-b430-524f36a88991">]
```

**Comment by elrayle** [comment]
@terrellt I added the save.  #aggregated_by and #in_collections now both work as expected.  Prior to saving, I saw something very interesting with #aggregated_by...

```
> col1.aggregated_by 
=> [#<ActiveFedora::Base id: "col-1">,
 #<ActiveFedora::Base id: "work-1">,
 #<ActiveFedora::Base id: "col-2">,
 #<ActiveFedora::Base id: "work-4">,
 #<ActiveFedora::Base id: "work-1">,
 #<ActiveFedora::Base id: "work-1">,
 #<ActiveFedora::Base id: "work-1">,
 #<ActiveFedora::Base id: "work-1">,
 #<ActiveFedora::Base id: "col-1">,
 #<ActiveFedora::Base id: "col-1">]
```

Clearly I have left over kruft in my repository.  But, how is it that those objects are returned by #aggregated_by when no relationship has been added for col2 except the in memory relationship with col1, both of which are brand new?

**Comment by terrellt** [comment]
@elrayle The only way that'd happen is if you have proxy objects in the repository with a nil proxyFor, I think?

**Comment by terrellt** [comment]
https://github.com/projecthydra-labs/hydra-works/compare/remove_filters is sitting waiting for this to be merged, and should fix up hydra-works the same way.

**Comment by mjgiarlo** [comment]
Waiting for the @elrayle / @terrellt discussion to wind down before pressing the green button.

**Comment by elrayle** [comment]
@terrellt, I suppose it could be the case that there are proxies with proxyFor=nil.  Is there an easy way to test for that?

Either way, @mjgiarlo feel free to merge.  The issue I am seeing with aggregated_by is not created by this PR.  It exists at master as well.  If it is a problem, a new issue should be created.

**Pull request merged**